File: app/src/main/java/com/github/premnirmal/ticker/portfolio/AddPositionActivity.java
package com.github.premnirmal.ticker.portfolio;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.github.premnirmal.ticker.BaseActivity;
import com.github.premnirmal.ticker.Injector;
import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.tickerwidget.R;

import javax.inject.Inject;

/**
 * Created by premnirmal on 7/3/15.
 */
public class AddPositionActivity extends BaseActivity {

    public static final String TICKER = "TICKER";

    @Inject
    IStocksProvider stocksProvider;
    protected String ticker;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.inject(this);
        setContentView(R.layout.activity_positions);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        ticker = getIntent().getStringExtra(TICKER);
        final TextView name = (TextView) findViewById(R.id.tickerName);
        name.setText(ticker);

        findViewById(R.id.doneButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onDoneClicked();
            }
        });

        findViewById(R.id.skipButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
    }

    protected void onDoneClicked() {
        final EditText sharesView = (EditText) findViewById(R.id.shares);
        final EditText priceView = (EditText) findViewById(R.id.price);
        final float price = Float.parseFloat(priceView.getText().toString());
        final float shares = Float.parseFloat(sharesView.getText().toString());
        stocksProvider.addPosition(ticker, shares, price);
        finish();
    }
}

File: app/src/main/kotlin/com/github/premnirmal/ticker/portfolio/AddPositionActivity.kt
package com.github.premnirmal.ticker.portfolio

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import com.github.premnirmal.ticker.AppPreferences
import com.github.premnirmal.ticker.base.BaseActivity
import com.github.premnirmal.ticker.components.InAppMessage
import com.github.premnirmal.ticker.components.Injector
import com.github.premnirmal.ticker.dismissKeyboard
import com.github.premnirmal.ticker.network.data.Holding
import com.github.premnirmal.tickerwidget.R
import com.github.premnirmal.tickerwidget.databinding.ActivityPositionsBinding
import com.github.premnirmal.tickerwidget.databinding.LayoutPositionHoldingBinding
import kotlinx.coroutines.launch
import java.text.NumberFormat

/**
 * Created by premnirmal on 2/25/16.
 */
class AddPositionActivity : BaseActivity<ActivityPositionsBinding>() {

  companion object {
    const val QUOTE = "QUOTE"
    const val TICKER = "TICKER"
  }

  internal lateinit var ticker: String
  override val simpleName: String = "AddPositionActivity"


  override fun onCreate(savedInstanceState: Bundle?) {
    Injector.appComponent.inject(this)
    super.onCreate(savedInstanceState)
    binding.toolbar.setNavigationOnClickListener {
      finish()
    }
    if (intent.hasExtra(TICKER) && intent.getStringExtra(TICKER) != null) {
      ticker = intent.getStringExtra(TICKER)!!
    } else {
      ticker = ""
      InAppMessage.showToast(this, R.string.error_symbol)
      finish()
      return
    }
    binding.tickerName.text = ticker

    binding.addButton.setOnClickListener { onAddClicked() }

    binding.positionsHolder.removeAllViews()
    val position = stocksProvider.getPosition(ticker)
    position?.let {
      for (holding in position.holdings) {
        addPositionView(holding)
      }
    }
    updateTotal()
  }

  private fun onAddClicked() {
    val sharesView = binding.shares
    val priceView = binding.price
    val priceText = priceView.text.toString()
    val sharesText = sharesView.text.toString()
    if (priceText.isNotEmpty() && sharesText.isNotEmpty()) {
      var price = 0f
      var shares = 0f
      var success = true
      try {
        val numberFormat: NumberFormat = NumberFormat.getNumberInstance()
        price = numberFormat.parse(priceText)!!
            .toFloat()
      } catch (e: NumberFormatException) {
        binding.priceInputLayout.error = getString(R.string.invalid_number)
        success = false
      }
      try {
        val numberFormat: NumberFormat = NumberFormat.getNumberInstance()
        shares = numberFormat.parse(sharesText)!!
            .toFloat()
      } catch (e: NumberFormatException) {
        binding.sharesInputLayout.error = getString(R.string.invalid_number)
        success = false
      }
      // Check for zero shares.
      if (success) {
        if (shares == 0.0f) {
          binding.sharesInputLayout.error = getString(R.string.invalid_number)
          success = false
        }
      }
      if (success) {
        binding.priceInputLayout.error = null
        binding.sharesInputLayout.error = null
        lifecycleScope.launch {
          val holding = stocksProvider.addHolding(ticker, shares, price)
          priceView.setText("")
          sharesView.setText("")
          addPositionView(holding)
          updateTotal()
          updateActivityResult()
        }
      }
    }
    dismissKeyboard()
  }

  private fun updateActivityResult() {
    val quote = checkNotNull(stocksProvider.getStock(ticker))
    val data = Intent()
    data.putExtra(QUOTE, quote)
    setResult(Activity.RESULT_OK, data)
  }

  private fun addPositionView(holding: Holding) {
    val positionBinding = LayoutPositionHoldingBinding.inflate(layoutInflater)
    val positionNumShares = positionBinding.positionShares
    val positionPrice = positionBinding.positionPrice
    val positionTotalValue = positionBinding.positionTotalValue
    positionNumShares.text = AppPreferences.DECIMAL_FORMAT.format(holding.shares)
    positionPrice.text = AppPreferences.DECIMAL_FORMAT.format(holding.price)
    positionTotalValue.text = AppPreferences.DECIMAL_FORMAT.format(holding.totalValue())
    binding.positionsHolder.addView(positionBinding.root)
    positionBinding.root.tag = holding
    // Remove entry when right side 'x' icon is clicked.
    positionBinding.removePosition.setOnClickListener {
      AlertDialog.Builder(this)
          .setTitle(R.string.remove)
          .setMessage(getString(R.string.remove_holding, "${holding.shares}@${holding.price}"))
          .setPositiveButton(R.string.remove) { dialog, _ ->
            lifecycleScope.launch {
              stocksProvider.removePosition(ticker, holding)
              binding.positionsHolder.removeView(positionBinding.root)
              updateTotal()
            }
            dialog.dismiss()
          }
          .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
          .show()
    }
  }

  private fun updateTotal() {
    val quote = checkNotNull(stocksProvider.getStock(ticker))
    binding.totalShares.text = quote.numSharesString()
    binding.averagePrice.text = quote.averagePositionPrice()
    binding.totalValue.text = quote.totalSpentString()
  }
}